### ISSUE TITLE
assert (!(p & 0xffff0000)); /*where p is pid*/

### ISSUE THREAD

**Issue by iqdoctor** [TelegramMessenger/MTProxy#41]
Hello, frends!

On my hosting when the application starts, it ends up on
`assert (! (p & 0xffff0000));`
in 
`common/pid.c:32`

if I comment on this line the application is launched, but all system connection calls are terminated with the `EINPROGRESS` return code and the process consumes ~ 70% of the CPU

Could you please clarify what this check makes `p & 0xffff0000`

Could you clarify what this check makes `p & 0xffff0000` and what rights in the system need to be checked, because on another hosting, everything works without any noticing.

And by the way, yes, the process pid> 110000, if this is important, but apparently important.

`iqdoctor @ raccoon4x: ~ / MTProxy $ grep connect st | grep 43
connect (43, {sa_family = AF_INET, sin_port = htons (8888), sin_addr = inet_addr ("91.108.4.135")}, 16) = -1 EINPROGRESS (Operation now in progress)
connect (43, {sa_family = AF_INET, sin_port = htons (80), sin_addr = inet_addr ("149.154.162.39")}, 16) = -1 EINPROGRESS (Operation now in progress)
connect (43, {sa_family = AF_INET, sin_port = htons (8888), sin_addr = inet_addr ("91.108.4.175")}, 16) = -1 EINPROGRESS (Operation now in progress)`

**Comment by iqdoctor** [comment]
Platform - Azure.

**Comment by stek29** [comment]
So, I've went through the source code and realized that:
pid_t on most platforms is int == 4 bytes.

```c
struct process_id {
unsigned ip;
short port;
unsigned short pid;
int utime;
};
```

As you can see, PID.pid is short == 2 bytes.
Doing `assert (! (p & 0xffff0000))` ensures that PID fits in those 2 bytes.

During handshakes process_id is passed around, and it's used to guard from starting handshake with one pid and finishing it with other one.

However, I don't think this would cause much trouble for mtproxies, since they don't face such a huge load as Telegram's mtfronts.

Also, from what I see, mtproxy only forks as much times as `-M` arg says, and that happens in pre_init.

So, the only problem I can see is that PIDs somehow change rapidly during those fork's, and the pid of one worker is the same as pid of another worker if you only consider last 2 bytes.

That means that 65536 processes should be spawned in between of those forks, which I consider very unlikely.

So, I think that this check could be omitted, and maybe `pids` after all forks should be checked that for any x and y in pids (x & 0xffff) != (y & 0xffff) where x != y.

**Comment by shadowlmd** [comment]
Temporary solution is to add `kernel.pid_max=65535` to /etc/sysctl.conf, that should be fine for most setups.

**Comment by Feliksas** [comment]
In Ubuntu Focal kernel.pid_max is raised to 4194304, this needs to be fixed ASAP.

**Comment by HugoFlorentino** [comment]
> 
> 
> In Ubuntu Focal kernel.pid_max is raised to 4194304, this needs to be fixed ASAP.

Yes, I have noticed that mtproto-proxy fails to launch but if you restart the OS, it launches without problems. Apparently the app receives a high PID when the OS has more than two days running.

What puzzles me is why does the service break, usually before 72h.

**Comment by ww898** [comment]
Is there any updates? Or the project was abandoned.

**Comment by ikigai-gh** [comment]
> Is there any updates? Or the project was abandoned.

Apparently, yes. I suggest just comment out the line with assertion

**Comment by heni** [comment]
> > Is there any updates? Or the project was abandoned.
> 
> Apparently, yes. I suggest just comment out the line with assertion

It wouldn't be enough to just remove assert. Internal npid_t structure should be changed also.

**Comment by heni** [comment]
One can check my simple fix for this issue - it works perfectly at my setup (Ubuntu 20.04.4 LTS in DO Cloud).
https://github.com/TelegramMessenger/MTProxy/pull/486

**Comment by AMajor-C** [comment]
> One can check my simple fix for this issue - it works perfectly at my setup (Ubuntu 20.04.4 LTS in DO Cloud). #486

Your fix makes mtproxy crazy, without clients it consumes 70% of core and actually doesnt work
// CentOS 8 Stream

**Comment by heni** [comment]
> > One can check my simple fix for this issue - it works perfectly at my setup (Ubuntu 20.04.4 LTS in DO Cloud). #486
> 
> Your fix makes mtproxy crazy, without clients it consumes 70% of core and actually doesnt work // CentOS 8 Stream

Could you please provide some logs and timings within the system setup to be able check your problem. I don't this behaviour on my setup - so it's hard to understand crazyness reason.

**Comment by D13410N3** [comment]
@heni btw the same for me - mtproxy is consuming much cpu and it doesn't work at all - proxy is unavailable, mtproxy doesn't bind to specified local port. No errors in the log

**Comment by YanceyChiew** [comment]
> ... - mtproxy is consuming much cpu ...

For reducing cpu time cusuming, you can try [my fork](https://github.com/YanceyChiew/MTProxy).

I've changed some logic about epoll timeout, so that if you specific timeout to -1 and using multithread mode, epoll would never timeout until an event be caught.

**Comment by igorechek06** [comment]
Why does this issue still open?

**Comment by AMajor-C** [comment]
probably because its still not fixed ?

**Comment by kr-ilya** [comment]
@levlam please pay attention to this issue it is still relevant

**Comment by levlam** [comment]
@kr-ilya The proxy doesn't support operating systems with non-16-bit process identifiers and there is no simple way to fix.

**Comment by va0816** [comment]
As manualy workaround possible to find empty pid, for example with 
ps aux
And run empty process with pid below 65535, for example 1233
echo 1233 | sudo tee /proc/sys/kernel/ns_last_pid &&  ./mtproto-proxy ...

**Comment by heni** [comment]
> As manualy workaround possible to find empty pid, for example with ps aux And run empty process with pid below 65535, for example 1233 echo 1233 | sudo tee /proc/sys/kernel/ns_last_pid && ./mtproto-proxy ...

You can just try changes from [this PR #486](https://github.com/TelegramMessenger/MTProxy/pull/486) 
I recently updated it - this patch perfectly compatible with master and works with pids > 65535 on my ubuntu 24.04 setup

**Comment by AlexLov** [comment]
I can confirm the patch from #486 PR works. Tested it on my nixos.
Just in case, here is nixos snippet to apply the patch for quick copy-paste: 

```nix
 mtproxy = pkgs.mtproxy.overrideAttrs (finalAttrs: prevAttrs: {
   patches = (prevAttrs.patches or []) ++ [
     (pkgs.fetchpatch {
       name = "long_pids";
       url = "https://patch-diff.githubusercontent.com/raw/TelegramMessenger/MTProxy/pull/486.patch";
       sha256 = "sha256-EboE6NPuwyFuvcMjX08B8Bxq57Gj1NmWhVXOBVyu+0s=";
     })
   ];
 });

```
